### PR TITLE
CenterCrop Layer

### PR DESCRIPTION
**Pull request by AdamLang96**
* Center Crop Preprocessing Layer Co-authored-by:
David Kim (@koyykdy) <dok098@ucsd.edu>
Brian Zheng (@brianzheng123) <brianzheng345@gmail.com>

- - -
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6875)

### CODE REVIEW

**Comment by AdamLang96** [comment]
@mattsoulanille Please take a look whenever you're available. Thanks!

**Comment by AdamLang96** [comment]
Hi @mattsoulanille,

Thanks again for meeting with us this morning! After reviewing the feedback on our last two layers, I am pretty confident we have some repeated mistakes in this one. @koyykdy and I will review this layer and ping you once we have cleaned those up to try and make things easier on your end, so don't worry about reviewing it for now.

Thanks,

Adam

**Comment by mattsoulanille** [comment]
Thanks for letting me know, @AdamLang96. I'll wait for your ping.

**Comment by AdamLang96** [comment]
@mattsoulanille I think this is ready for review whenever you're available. Thank you!

File: tfjs-layers/src/layers/preprocessing/center_crop.ts [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    } else {
+      input = inputs as Tensor4D;
+    }
+
+    for (let i = 0; i < input.shape[0]; i++) {
+      boxesArr.push(bound);
+    }
+
+    const boxes: Tensor2D  = tensor(boxesArr, [boxesArr.length, 4]);
+    const boxInd: Tensor1D = tensor([...Array(boxesArr.length).keys()],
```
Comment (mattsoulanille) [inline tfjs-layers/src/layers/preprocessing/center_crop.ts:51]: Use [`tf.range`](https://js.tensorflow.org/api/latest/#range) instead.

**Review by mattsoulanille — CHANGES_REQUESTED** [review]
The logic looks good. I just have a few style changes. Thanks!

File: tfjs-layers/src/layers/preprocessing/center_crop.ts [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+  call(inputs: Tensor3D | Tensor4D , kwargs: Kwargs): Tensor[] | Tensor {
+    const rankedInputs = getExactlyOneTensor(inputs) as Tensor3D | Tensor4D;
+    const dtype = rankedInputs.dtype;
+    const hAxis = rankedInputs.shape.length - 3;
+    const wAxis = rankedInputs.shape.length - 2;
+    const inputShape = rankedInputs.shape;
+    const inputHeight = inputShape[hAxis];
+    const inputWidth = inputShape[wAxis];
```
[the remark this answers is not in the document]
Comment (AdamLang96) [reply tfjs-layers/src/layers/preprocessing/center_crop.ts]: @mattsoulanille I may be wrong but I believe `Array.at()` is only available for TypeScript `4.6>` and `tfjs` is currently running `3.5.3`

**Comment by AdamLang96** [comment]
@mattsoulanille I updated the layer to reflect your changes. Let me know what you think. Thanks!

**Review by mattsoulanille — APPROVED** [review]
LGTM. Thanks!

File: tfjs-layers/src/layers/preprocessing/center_crop.ts [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    this.width = args.width;
+  }
+
+  centerCrop(inputs: Tensor3D | Tensor4D, hBuffer: number, wBuffer: number,
+            height: number, width: number, inputHeight: number,
+            inputWidth: number, dtype: DataType): Tensor | Tensor[] {
+
+    return tidy(() => {
+      let input: Tensor4D;
+      let rank3      = false;
```
Comment (ahmedsabie) [inline tfjs-layers/src/layers/preprocessing/center_crop.ts:42]: Can you rename to isRank3

**Review by ahmedsabie — CHANGES_REQUESTED** [review]

File: tfjs-layers/src/layers/preprocessing/center_crop.ts (hunk 3 above)
Comment (AdamLang96) [reply tfjs-layers/src/layers/preprocessing/center_crop.ts]: @ahmedsabie done!

**Comment by mattsoulanille** [comment]
This is failing `tslint`:
```
$ tslint -p tsconfig_tslint.json
/workspace/tfjs-layers/src/exports_layers.ts:1770:1
ERROR: 1770:1  jsdoc-format  asterisks in jsdoc must be aligned
ERROR: 1771:1  jsdoc-format  asterisks in jsdoc must be aligned
ERROR: 1772:1  jsdoc-format  asterisks in jsdoc must be aligned
ERROR: 1773:1  jsdoc-format  asterisks in jsdoc must be aligned
ERROR: 1774:1  jsdoc-format  asterisks in jsdoc must be aligned
ERROR: 1775:1  jsdoc-format  asterisks in jsdoc must be aligned
ERROR: 1776:1  jsdoc-format  asterisks in jsdoc must be aligned
ERROR: 1777:1  jsdoc-format  asterisks in jsdoc must be aligned
ERROR: 1778:1  jsdoc-format  asterisks in jsdoc must be aligned
ERROR: 1779:1  jsdoc-format  asterisks in jsdoc must be aligned
ERROR: 1780:1  jsdoc-format  asterisks in jsdoc must be aligned
ERROR: 1781:1  jsdoc-format  asterisks in jsdoc must be aligned
ERROR: 1782:1  jsdoc-format  asterisks in jsdoc must be aligned
ERROR: 1783:1  jsdoc-format  asterisks in jsdoc must be aligned
ERROR: 1784:1  jsdoc-format  asterisks in jsdoc must be aligned
ERROR: 1785:1  jsdoc-format  asterisks in jsdoc must be aligned
ERROR: 1786:1  jsdoc-format  asterisks in jsdoc must be aligned
ERROR: 1787:1  jsdoc-format  asterisks in jsdoc must be aligned
ERROR: 1788:1  jsdoc-format  asterisks in jsdoc must be aligned
ERROR: 1789:1  jsdoc-format  asterisks in jsdoc must be aligned
ERROR: 1790:1  jsdoc-format  asterisks in jsdoc must be aligned
```

**Comment by AdamLang96** [comment]
@mattsoulanille whoops just fixed it

**Comment by mattsoulanille** [comment]
Thanks!
@ahmedsabie Please take another look.

**Review by ahmedsabie — APPROVED** [review]

**Pull request merged**